zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+    use frame_support::{
+        ensure,
+        pallet_prelude::{
+            DispatchResultWithPostInfo, OptionQuery, StorageDoubleMap, StorageMap, ValueQuery,
+        },
+        traits::{Currency, Get, Hooks, IsType, LockIdentifier, LockableCurrency, WithdrawReasons},
+        Blake2_128Concat, BoundedVec, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use sp_runtime::{
+        traits::{One, Saturating, Zero},
+        DispatchError, SaturatedConversion,
+    };
+    use zeitgeist_primitives::types::{MarketStatus, OutcomeReport};
+
+    pub(crate) type BalanceOf<T> =
+        <CurrencyOf<T> as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+    pub(crate) type CurrencyOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::Currency;
+    pub(crate) type MarketIdOf<T> =
+        <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Votes on a dispute after there are already two disputes and the 'DisputePeriod' is not over.
+        /// NOTE: In the 'DisputePeriod' voting on a dispute is allowed.
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::vote_on_outcome())]
+        pub fn vote_on_outcome(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            outcome_index: u32,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin)?;
+            ensure!(
+                amount <= CurrencyOf::<T>::free_balance(&sender),
+                Error::<T>::InsufficientAmount
+            );
+            ensure!(amount >= T::MinDisputeVoteAmount::get(), Error::<T>::AmountTooLow);
+
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Disputed, Error::<T>::InvalidMarketStatus);
+
+            ensure!(
+                <Outcomes<T>>::get(market_id).len() >= T::MinOutcomes::get() as usize,
+                Error::<T>::NotEnoughOutcomes
+            );
+
+            // dispute vote is already present because of the dispute bond of the disputor
+            let vote_balance = <OutcomeVotes<T>>::get(market_id, outcome_index)
+                .ok_or(Error::<T>::OutcomeDoesNotExist)?;
+
+            <LockInfoOf<T>>::mutate(&sender, market_id, |locked_balance| {
+                *locked_balance = Some(locked_balance.map_or(amount, |x| x.max(amount)));
+            });
+
+            CurrencyOf::<T>::extend_lock(
+                T::VoteLockIdentifier::get(),
+                &sender,
+                amount,
+                WithdrawReasons::TRANSFER,
+            );
+
+            <OutcomeVotes<T>>::insert(
+                market_id,
+                outcome_index,
+                vote_balance.saturating_add(amount),
+            );
```

Comment:
At the moment it's possible, that a voter can reduce the votes amount overall for an outcome. It's up to you, if you think this is unwanted behaviour. I don't feel this is a big problem, because it's a feature, not a bug to reduce the relevance of an outcome. 

How is that possible? Assume you have four outcomes A, B, C, D. Now some voter could vote with the exact same `locked_balance` on B, C, D to reduce the relevance of A. For me, it's a cool feature. How do you think about this? I am open to discuss.

Reply:
Major bug at the moment that needs to get fixed: vote_on_outcome allows to increase the outcomes balance multiple times with the same locked tokens.